shared/lib/estimators/kNeighborsRegressor.ts

```diff
@@ -0,0 +1,175 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { Scikit1D, Scikit2D } from '../types'
+import {
+  KNeighborsBase,
+  KNeighborsBaseParams
+} from '../neighbors/kNeighborsBase'
+import { convertToNumericTensor1D_2D } from '../utils'
+import { Scalar, Tensor1D } from '@tensorflow/tfjs'
+import { tf } from '../../globals'
+
+type int = number
+type WeightsFn = (distances: number[]) => ArrayLike<number>
+
+const WEIGHTS_FUNCTIONS = {
+  uniform(distances: number[]) {
+    distances.fill(1 / distances.length)
+    return distances
+  },
+  distance(distances: number[]) {
+    const len = distances.length
+
+    const nInf = distances.reduce((nInf, x) => nInf + +(x === Infinity), 0)
+
+    if (nInf === 0) distances.fill(1 / len)
+    else
+      for (let i = len; i-- > 0; )
+        distances[i] = +(distances[i] === Infinity) / nInf
+
+    return distances
+  }
+  //  softmax( distances: number[] )
+  //  {
+  //    const len = distances.length
+  //
+  //    const min = distances.reduce( (x, y) => Math.min(x, y) )
+  //
+  //    let sum = 0
+  //    for (let i = len; i-- > 0;) {
+  //      const di = Math.exp(min - distances[i])
+  //      sum += di
+  //      distances[i] = di
+  //    }
+  //
+  //    for (let i = len; i-- > 0;)
+  //      distances[i] /= sum
+  //
+  //    return distances
+  //  }
+}
+
+/**
+ * Wraps a weightFn with input and ouput checks.
+ *
+ * @param weightsFn The wrapped weight function.
+ * @returns `weightsFn` wrapped with checks/assertions.
+ */
+const checkedWeightsFn = (weightsFn: WeightsFn) => (distances: number[]) => {
+  if (distances.some((d) => isNaN(d) || d < 0))
+    throw new Error(
+      'KNeighborsRegressor: metric function returned invalid distance.'
+    )
+
+  const weights = weightsFn(distances)
+
+  if (weights.length !== distances.length)
+    throw new Error(
+      'KNeighborsRegressor: array returned by weights must have the same length than the input.'
+    )
+
+  let sum = 0
+  for (let i = weights.length; i-- > 0; ) {
+    const wi = weights[i]
+    if (!(0 <= wi && wi <= 1))
+      // <- handles NaN
+      throw new Error(
+        'KNeighborsRegressor: entries returned by weights must be in range [0,1].'
+      )
+    sum += wi
+  }
+  if (!(Math.abs(sum - 1) <= 1e-8))
+    // <- handles NaN
+    throw new Error(
+      'KNeighborsRegressor: entries returned by weights must sum up to 1.'
+    )
+
+  return weights
+}
+
+export interface KNeighborsRegressorParams extends KNeighborsBaseParams {
+  nNeighbors?: int
+  weights?: WeightsFn | keyof typeof WEIGHTS_FUNCTIONS
+}
+
+export class KNeighborsRegressor extends KNeighborsBase {
+  private _weightsFn: WeightsFn
+  private _nNeighbors: int
```

Comment:
Is there a good reason why these are `private`? Doesn't that mean that after a user creates a `KNeighborsRegressor`, they can't simple print what the value of k is (the number of neighbors?

```
let knn = new KNeighborsRegressor()
console.log(knn._nNeighbors) // illegal
```

Reply:
Good point, `nNeighbors` should be public. `weightsFn` is an implementation detail though. But I can also add `weights` as property.